tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -0,0 +1,518 @@
+			a.resourceTrackerDir, err)
+	}
+	log.Info("Successfully created windows instance: %s, please RDP into windows with the following:")
+	log.Info("xfreerdp /u:Administrator /v:%s  /h:1080 /w:1920 /p:'Secret2018'", ipAddr)
+	return nil
+}
+
+// DestroyWindowsVM destroys the created instance and security group on AWS specified in 'windows-node-installer.json'
+// file.
+func (a *awsProvider) DestroyWindowsVM() error {
+	// Read from `windows-node-installer.json` file.
+	instances, err := resource.ReadInstallerInfo(a.resourceTrackerDir)
+	log.Info("consuming file '%v'", a.resourceTrackerDir)
+	if err != nil {
+		return err
+	}
+	// Delete all instances from the json file.
+	for _, instanceID := range instances.InstanceIDs {
+		err = a.terminateInstance(instanceID)
+		if err != nil {
+			log.Info("failed to terminate instance: %s", instanceID)
+		}
+	}
+	// Wait for instances termination after they are initiated.
+	for _, instanceID := range instances.InstanceIDs {
+		err = a.waitUntilInstanceTerminated(instanceID)
+		if err != nil {
+			log.Info("failed to wait for instance: %s to terminate", instanceID)
+		}
+	}
+	// Delete security groups after associated instances are terminated.
+	for _, sgID := range instances.SecurityGroupIDs {
+		sgInUse, err := a.isSGInUse(sgID)
+		if err != nil {
+			log.Info("failed to check if security group %s is in use, %v", sgID, err)
+		}
+		if !sgInUse {
+			err = a.deleteSG(sgID)
+			if err != nil {
+				log.Info("failed to delete Security Group: %s", sgID)
+			}
+		}
+	}
+	err = resource.RemoveInstallerInfo(instances, a.resourceTrackerDir)
+	if err != nil {
+		log.Info("%s file is not updated, %v", resource.InstallerInfoFileName, err)
+	}
+	return nil
+}
+
+// createInstances creates one VM instance based on the given information and returns if no instance is created.
+func (a *awsProvider) createInstances(imageID, instanceType, sshKey, subnetID, sgID, workerSgID string,
+	iamProfile *ec2.IamInstanceProfileSpecification) (string, error) {
+	runResult, err := a.EC2.RunInstances(&ec2.RunInstancesInput{
+		ImageId:            aws.String(imageID),
+		InstanceType:       aws.String(instanceType),
+		KeyName:            aws.String(sshKey),
+		SubnetId:           aws.String(subnetID),
+		MinCount:           aws.Int64(1),
+		MaxCount:           aws.Int64(1),
+		IamInstanceProfile: iamProfile,
+		SecurityGroupIds:   aws.StringSlice([]string{sgID, workerSgID}),
+	})
+	if err != nil {
+		return "", err
+	}
+	if len(runResult.Instances) < 1 {
+		return "", fmt.Errorf("failed to create an instance")
+	}
+	return *runResult.Instances[0].InstanceId, nil
+}
+
+// findOrCreateSg tries to find a Security Group based on VPC ID and Group Name. If no such Security Group exists,
+// it creates a Security Group under the VPC with a Group Name '<infraID>-sg'.
+// The function returns Security Group ID or error for both finding and creating a Security Group.
+func (a *awsProvider) findOrCreateSg(infraID, vpcID string) (string, error) {
```

Comment:
You shouldn't create a security group if you can't find one. Just return an error and not proceed further in the call site. Remember the intent is to create a Windows VM that can become a worker node and not just create a Windows VM.